libs/langchain-community/src/chat_models/watsonx_ai.ts

```diff
@@ -0,0 +1,157 @@
+import {
+  AIMessageChunk,
+  BaseMessage,
+  ChatMessage,
+} from "@langchain/core/messages";
+import { ChatGenerationChunk } from "@langchain/core/outputs";
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import type {
+  WatsonModelParameters,
+  WatsonxAIParams,
+} from "../types/watsonx-types.js";
+import { WatsonApiClient } from "../utils/watsonx-client.js";
+
+export class WatsonxAIChat extends SimpleChatModel {
+  private readonly watsonApiClient: WatsonApiClient;
+
+  readonly modelId!: string;
+
+  readonly modelParameters?: WatsonModelParameters;
+
+  readonly projectId!: string;
+
+  constructor(fields: WatsonxAIParams & BaseChatModelParams) {
+    super(fields);
+
+    const {
+      clientConfig = {},
+      modelId = "meta-llama/llama-2-70b-chat",
+      modelParameters,
+      projectId = getEnvironmentVariable("WATSONX_PROJECT_ID") ?? "",
+    } = fields;
+
+    this.modelId = modelId;
+    this.modelParameters = modelParameters;
+    this.projectId = projectId;
+
+    const {
+      apiKey = getEnvironmentVariable("IBM_CLOUD_API_KEY"),
+      apiVersion = "2023-05-29",
+      region = "us-south",
+    } = clientConfig;
+
+    if (!apiKey) {
+      throw new Error("Missing IBM Cloud API Key");
+    }
+
+    if (!this.projectId) {
+      throw new Error("Missing WatsonX AI Project ID");
+    }
+
+    this.watsonApiClient = new WatsonApiClient({
+      apiKey,
+      apiVersion,
+      region,
+    });
+  }
+
+  protected _formatMessagesAsPrompt(messages: BaseMessage[]): string {
+    return messages
+      .map((message) => {
+        let messageText;
+        if (message._getType() === "human") {
+          messageText = `[INST] ${message.content} [/INST]`;
+        } else if (message._getType() === "ai") {
+          messageText = message.content;
+        } else if (message._getType() === "system") {
+          messageText = `<<SYS>> ${message.content} <</SYS>>`;
+        } else if (ChatMessage.isInstance(message)) {
+          messageText = `\n\n${message.role[0].toUpperCase()}${message.role.slice(
+            1
+          )}: ${message.content}`;
+        } else {
+          console.warn(
+            `Unsupported message type passed to Watson: "${message._getType()}"`
+          );
+          messageText = "";
+        }
+        return messageText;
+      })
+      .join("\n");
+  }
+
+  _combineLLMOutput() {
+    return {};
+  }
+
+  async _call(
+    messages: BaseMessage[],
+    options: this["ParsedCallOptions"],
+    runManager: CallbackManagerForLLMRun | undefined
+  ): Promise<string> {
+    const chunks = [];
+    const stream = this._streamResponseChunks(messages, options, runManager);
+    for await (const chunk of stream) {
+      chunks.push(chunk.message.content);
+    }
+    return chunks.join("");
+  }
+
+  override async *_streamResponseChunks(
+    _messages: BaseMessage[],
+    _options: this["ParsedCallOptions"],
+    _runManager?: CallbackManagerForLLMRun
```

Comment:
Drop the underscore if they're being used. Typically, variables prefixed with an underscore are unused, and the underscore is used to bypass a lint rule for no-unused-variables